contracts/clients/CkbProof.sol

```diff
@@ -0,0 +1,288 @@
+    }
+
+    // Decode the content
+    bytes memory content = ls[2].toBytes();
+
+    // Return the decoded Envelope
+    return Envelope(msg_type, commitments, content);
+}
+
+function parseCommitment(
+    bytes memory ckb_transaction
+) pure returns (CommitmentKV[] memory) {
+    uint256 witness_count = ckb_transaction.readCKBTxWitnessCount();
+    uint8 output_type_index = 2;
+    (uint256 offset, uint256 size) = ckb_transaction.readCKBTxWitness(
+        uint8(witness_count),
+        output_type_index
+    );
+    bytes memory output_type_bytes = new bytes(size);
+    for (uint i = 0; i < size; i++) {
+        output_type_bytes[i] = ckb_transaction[i + offset];
+    }
+    Envelope memory witness_struct = decodeRlpEnvelope(output_type_bytes);
+    return witness_struct.commitments;
+}
+
+function verify_hash_exist(
+    bytes32[] memory leaves,
+    uint8 verify_type,
+    bytes32 transaction_hash,
+    bytes32 witness_hash
+) pure returns (bool) {
+    bytes32 hash;
+    if (verify_type == 0) {
+        hash = transaction_hash;
+    } else {
+        hash = witness_hash;
+    }
+
+    bool isInLeaves = false;
+    for (uint i = 0; i < leaves.length; i++) {
+        if (leaves[i] == hash) {
+            isInLeaves = true;
+            break;
+        }
+    }
+
+    return isInLeaves;
+}
+
+function isCommitInCommitments(
+    CommitmentKV[] memory commitments,
+    bytes memory key,
+    bytes calldata value
+) pure returns (bool) {
+    uint256 key_hash = uint256(keccak256(key));
+    uint256 value_hash = uint256(keccak256(value));
+    for (uint i = 0; i < commitments.length; i++) {
+        if (
+            commitments[i].key == key_hash && commitments[i].value == value_hash
+        ) {
+            return true;
+        }
+    }
+    return false;
+}
+
+function verify_proof(
+    bytes memory abiEncodedProof,
+    bytes memory path,
+    bytes calldata value
+) returns (bool) {
+    // Parse the proof from the abi encoded data
+    AxonObjectProof memory axon_obj_proof = parseProof(abiEncodedProof);
+
+    // Get the CKB header
+    CkbLightClient lightClient;
+    CKBHeader memory header = lightClient.getHeader(axon_obj_proof.block_hash);
+
+    // Calculate the transaction hash and witness hash
+    (bytes32 transaction_hash, bytes32 witness_hash) = calculateHashes(
+        axon_obj_proof.ckb_transaction
+    );
+
+    // Check if the transaction hash or witness hash is in the leaves
+    if (
+        !verify_hash_exist(
+            axon_obj_proof.proof_payload.proof.leaves,
+            axon_obj_proof.proof_payload.verify_type,
+            transaction_hash,
+            witness_hash
+        )
+    ) {
+        return false;
+    }
+
+    // Create the VerifyProofPayload
+    VerifyProofPayload memory payload = VerifyProofPayload({
+        verify_type: axon_obj_proof.proof_payload.verify_type,
+        transactions_root: header.transactionsRoot,
+        witnesses_root: witness_hash,
```

Comment:
This doesn't make sense to me: why would `witnesses_root` be the witness hash of the transaction?

Reply:
These should be provided by the relayer.

Reply:
ok, will change this to 
     witnessesRoot: axon_obj_proof.proofPayload.witnessesRoot,
     rawTransactionsRoot: axon_obj_proof.proofPayload.rawTransactionsRoot,